### PR TITLE
52 cronapi

### PR DESCRIPTION
**Pull request by dchertkov**

### CODE REVIEW

File: packages/api/contract.go [hunk 1]
```diff
@@ -52,9 +52,16 @@ func contract(w http.ResponseWriter, r *http.Request, data *apiData, logger *log
 	}
 	info := (*contract).Block.Info.(*script.ContractInfo)
 
+	var signedBy int64
```
Comment (potehinre) [inline packages/api/contract.go:55]: seems like this variable is useless, signID is enough

File: packages/api/contract.go [hunk 2]
```diff
@@ -52,9 +52,16 @@ func contract(w http.ResponseWriter, r *http.Request, data *apiData, logger *log
...
+	signID := data.keyId
+	if data.params[`signed_by`] != nil {
+		signedBy = data.params[`signed_by`].(int64)
```
Comment (potehinre) [inline packages/api/contract.go:58]: signID = data.params["signed_by"].(int64)

File: packages/api/contract.go [hunk 3]
```diff
@@ -129,6 +136,7 @@ func contract(w http.ResponseWriter, r *http.Request, data *apiData, logger *log
 		TokenEcosystem: data.params[`token_ecosystem`].(int64),
 		MaxSum:         data.params[`max_sum`].(string),
 		PayOver:        data.params[`payover`].(string),
+		SignedBy:       signedBy,
```
Comment (potehinre) [inline packages/api/contract.go:139]: signID

File: packages/api/node.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	pubkey, err := hex.DecodeString(NodePublicKey)
+	if err != nil {
+		logger.WithFields(log.Fields{"type": consts.ConversionError, "error": err}).Error("decoding private key from hex")
+		return err
+	}
+	data.params[`signed_by`] = smart.PubToID(NodePublicKey)
+	prepareData := *data
+	if err = prepareContract(w, r, &prepareData, logger); err != nil {
+		logger.WithFields(log.Fields{"type": consts.APIError}).Error("can't prepare contract")
```
Comment (potehinre) [inline packages/api/node.go]: Unnecessary logging

File: packages/api/node.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		return err
+	}
+	signed, err := crypto.Sign(NodePrivateKey, prepareData.result.(prepareResult).ForSign)
```
Comment (potehinre) [inline packages/api/node.go]: Need to log error

File: packages/api/node.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	data.params[`signature`] = signed
+	data.params[`pubkey`] = pubkey
+	data.params[`time`] = prepareData.result.(prepareResult).Time
+	if err = contract(w, r, data, logger); err != nil {
+		logger.WithFields(log.Fields{"type": consts.APIError}).Error("can't call contract")
```
Comment (potehinre) [inline packages/api/node.go]: Unnecessary logging

File: packages/api/vde.go [hunk 7]
```diff
@@ -157,23 +158,33 @@ func InitSmartContract(sc *smart.SmartContract, data []byte) error {
 }
 
 // VDEContract is init VDE contract
-func VDEContract(data []byte) (result *contractResult, err error) {
+func VDEContract(contractData []byte, data *apiData) (result *contractResult, err error) {
 	var ret string
-	hash, err := crypto.Hash(data)
+	hash, err := crypto.Hash(contractData)
 	if err != nil {
 		return
```
Comment (potehinre) [inline packages/api/vde.go]: Need to log CryptoError

File: packages/api/vde_test.go [hunk 8]
```diff
@@ -490,3 +493,164 @@ func TestHTTPRequest(t *testing.T) {
 		return
 	}
 }
+
+func TestNodeHTTPRequest(t *testing.T) {
```
Comment (potehinre) [inline packages/api/vde_test.go:497]: Need to port whole test to testify

File: packages/consts/log_types.go [hunk 9]
```diff
@@ -49,4 +49,6 @@ const (
 	StatsdError              = "StatsdError"
 	MigrationError           = "MigrationError"
 	AutoupdateError          = "AutoupdateError"
+	APIError                 = "APIError"
```
Comment (potehinre) [inline packages/consts/log_types.go]: Unnecessary errors

File: packages/daemons/scheduler.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	for _, cronTask := range tasks {
+		err = scheduler.UpdateTask(&scheduler.Task{
+			ID:       cronTask.ID,
+			CronSpec: cronTask.Cron,
+			Handler: &contract.ContractHandler{
+				Contract: cronTask.Contract,
+			},
+		})
+		if err != nil {
+			log.WithFields(log.Fields{"type": consts.SchedulerError, "error": err}).Error("update cron task")
```
Comment (potehinre) [inline packages/daemons/scheduler.go]: Unnecessary logging

File: packages/scheduler/contract/request.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+func NodeContract(Name string) (result contractResult, err error) {
+	var (
+		sign                          []byte
+		ret                           authResult
+		NodePrivateKey, NodePublicKey string
+	)
+	err = sendAPIRequest(`GET`, `getuid`, nil, &ret, ``)
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.APIError, "error": err}).Error("calling getuid")
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: Unnecessary logging

File: packages/scheduler/contract/request.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+		return
+	}
+	auth := ret.Token
+	if len(ret.UID) == 0 {
+		err = fmt.Errorf(`getuid has returned empty uid`)
+		log.WithFields(log.Fields{"type": consts.APIError, "error": err}).Error("empty uid")
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: Unnecessary logging

File: packages/scheduler/contract/request.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.CryptoError, "error": err}).Error("signing node uid")
+		return
+	}
+	form := url.Values{"pubkey": {NodePublicKey}, "signature": {hex.EncodeToString(sign)},
+		`ecosystem`: {converter.Int64ToStr(1)}}
+	var logret authResult
+	err = sendAPIRequest(`POST`, `login`, &form, &logret, auth)
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.APIError, "error": err}).Error("login node")
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: Unnecessary logging

File: packages/scheduler/contract/request.go [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+		return
+	}
+	auth = logret.Token
+	form = url.Values{`vde`: {`true`}}
+	err = sendAPIRequest(`POST`, `node/`+Name, &form, &result, auth)
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.APIError, "error": err}).Error("node request")
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: Unnecessary logging

File: packages/scheduler/contract/request.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+func sendAPIRequest(rtype, url string, form *url.Values, v interface{}, auth string) error {
+	client := &http.Client{}
+	var ioform io.Reader
+	if form != nil {
+		ioform = strings.NewReader(form.Encode())
+	}
+	req, err := http.NewRequest(rtype, fmt.Sprintf(`http://%s:%d%s%s`, conf.Config.HTTP.Host,
+		conf.Config.HTTP.Port, consts.ApiPath, url), ioform)
+	if err != nil {
+		return err
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: logging needed

File: packages/scheduler/contract/request.go [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	if len(auth) > 0 {
+		req.Header.Set("Authorization", headerAuthPrefix+auth)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: logging needed

File: packages/scheduler/contract/request.go [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	defer resp.Body.Close()
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: Logging needed

File: packages/scheduler/contract/request.go [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf(`%d %s`, resp.StatusCode, strings.TrimSpace(string(data)))
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: logging needed

File: packages/scheduler/contract/request.go [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	if err = json.Unmarshal(data, v); err != nil {
+		return err
```
Comment (potehinre) [inline packages/scheduler/contract/request.go]: logging needed

File: packages/scheduler/scheduler_test.go [hunk 20]
```diff
@@ -0,0 +1,76 @@
+package scheduler
```
Comment (potehinre) [inline packages/scheduler/scheduler_test.go:1]: need port to testify

File: packages/scheduler/task.go [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+	return fmt.Sprintf("id %d cron %s", t.ID, t.CronSpec)
+}
+
+func (t *Task) ParseCron() error {
+	if len(t.CronSpec) == 0 {
+		return nil
+	}
+
+	var err error
+	t.schedule, err = cron.Parse(t.CronSpec)
```
Comment (potehinre) [inline packages/scheduler/task.go]: need error logging

File: packages/smart/funcs.go [hunk 22]
```diff
@@ -1108,3 +1112,41 @@ func HTTPPostJSON(requrl string, headers map[string]interface{}, json_str string
 	}
 	return string(data), nil
 }
+
+func ValidateCron(cronSpec string) error {
+	err := scheduler.Parse(cronSpec)
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.ParseError, "error": err}).Error("validate cron format")
```
Comment (potehinre) [inline packages/smart/funcs.go]: unnecessary logging

File: packages/smart/funcs.go [hunk 23]
```diff
@@ -1108,3 +1112,41 @@ func HTTPPostJSON(requrl string, headers map[string]interface{}, json_str string
...
+
+	err = scheduler.UpdateTask(&scheduler.Task{
+		ID:       cronTask.ID,
+		CronSpec: cronTask.Cron,
+		Handler: &contract.ContractHandler{
+			Contract: cronTask.Contract,
+		},
+	})
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.SchedulerError, "error": err}).Error("update cron task")
```
Comment (potehinre) [inline packages/smart/funcs.go]: Unnecessary logging

File: packages/scheduler/scheduler.go [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+type Scheduler struct {
+	cron *cron.Cron
+}
+
+func (s *Scheduler) AddTask(t *Task) error {
+	err := t.ParseCron()
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.ParseError, "error": err}).Error("parse cron format")
```
Comment (potehinre) [inline packages/scheduler/scheduler.go]: Unnecessary logging

**Review by potehinre — CHANGES_REQUESTED** [review]

File: packages/api/contract.go (hunk 1 above)
Comment (gentee) [reply packages/api/contract.go:55]: signedBy can be params[`signed_by`] or zero.
signID  can be params[`signed_by`] or data.keyID.
They are the same only when params[`signed_by`] is defined.

File: packages/api/node.go (hunk 4 above)
Comment (gentee) [reply packages/api/node.go]: Logging has been fixed.

File: packages/api/node.go [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+
+	log "github.com/sirupsen/logrus"
+)
+
+func nodeContract(w http.ResponseWriter, r *http.Request, data *apiData, logger *log.Entry) error {
+	var err error
+
+	NodePrivateKey, NodePublicKey, err := utils.GetNodeKeys()
+	if err != nil || len(NodePrivateKey) == 0 {
+		if err == nil {
```
Comment (ghost) [inline packages/api/node.go]: reduce nesting please
```
if err != nil {
    return err
}

if len(...) == 0 {
    log...
    return err
}
```

File: packages/api/node.go [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+	if err != nil {
+		logger.WithFields(log.Fields{"type": consts.ConversionError, "error": err}).Error("decoding private key from hex")
+		return err
+	}
+	data.params[`signed_by`] = smart.PubToID(NodePublicKey)
+	prepareData := *data
+	if err = prepareContract(w, r, &prepareData, logger); err != nil {
+		return err
+	}
+	signed, err := crypto.Sign(NodePrivateKey, prepareData.result.(prepareResult).ForSign)
```
Comment (ghost) [inline packages/api/node.go]: use more semantic names please. "signed" what?

File: packages/api/vde.go [hunk 27]
```diff
@@ -157,23 +158,34 @@ func InitSmartContract(sc *smart.SmartContract, data []byte) error {
...
+		log.WithFields(log.Fields{"type": consts.CryptoError, "error": err}).Error("getting hash of contract data")
 		return
 	}
 	result = &contractResult{Hash: hex.EncodeToString(hash)}
 
 	sc := smart.SmartContract{VDE: true, TxHash: hash}
-	err = InitSmartContract(&sc, data)
+	err = InitSmartContract(&sc, contractData)
 	if err == nil {
+		if data.token != nil && data.token.Valid {
```
Comment (ghost) [inline packages/api/vde.go]: reduce nesting, invert logic 
```
if err != nil {
    return ...
}

if data.token != nil .....
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by potehinre — APPROVED** [review]

File: packages/api/node.go (hunk 25 above)
Comment (gentee) [reply packages/api/node.go]: Fixed

File: packages/api/vde.go (hunk 27 above)
Comment (gentee) [reply packages/api/vde.go]: Fixed

**Review by ghost — APPROVED** [review]

**Pull request merged**